### PR TITLE
Clean up applied manifestworks when detaching cluster

### PR DESCRIPTION
**Pull request by elgnay**
Related issues:
- https://github.com/open-cluster-management/backlog/issues/6853
- https://github.com/open-cluster-management/backlog/issues/7312

Signed-off-by: Yang Le <yangle@redhat.com>

### CODE REVIEW

File: pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go [hunk 1]
```diff
@@ -356,12 +361,27 @@ func (n *klusterletController) cleanUp(ctx context.Context, controllerContext fa
 		return err
 	}
 
-	// Remove secret
-	err = n.kubeClient.CoreV1().Secrets(config.KlusterletNamespace).Delete(ctx, config.HubKubeConfigSecret, metav1.DeleteOptions{})
-	if err != nil && !errors.IsNotFound(err) {
+	// get hub host from secret and then remove the secret
+	var hubHost string
+	hubKubeConfigSecret, err := n.kubeClient.CoreV1().Secrets(config.KlusterletNamespace).Get(ctx, config.HubKubeConfigSecret, metav1.GetOptions{})
```
Comment (qiujian16) [inline pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go]: it needs more doc here.  And could we use bootstrap secret here?

File: pkg/helpers/helpers.go [hunk 2]
```diff
@@ -416,3 +418,37 @@ func UpdateKlusterletGenerationsFn(generations ...operatorapiv1.GenerationStatus
 		return nil
 	}
 }
+
+// LoadClientConfigFromSecret returns a client config loaded from the given secret
+func LoadClientConfigFromSecret(secret *corev1.Secret) (*restclient.Config, error) {
```
Comment (qiujian16) [inline pkg/helpers/helpers.go:423]: this is similar as the func here https://github.com/open-cluster-management/registration-operator/blob/master/pkg/operators/klusterlet/controllers/statuscontroller/klusterlet_status_controller.go#L287. Maybe do some refactor also

File: pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go [hunk 3]
```diff
@@ -427,6 +458,36 @@ func (n *klusterletController) removeKlusterletFinalizer(ctx context.Context, de
 	return nil
 }
 
+// cleanUpAppliedManifestWorks removes finalizer from AppliedManifestWorks which meet the
+// conditions below:
+//   1. The name of AppliedManifestWork starts with hash of the given hub host;
+//   2. The AppliedManifestWork is in terminating state;
```
Comment (qiujian16) [inline pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go]: why we only handle terminating resource?

File: pkg/helpers/helpers.go (hunk 2 above)
Comment (qiujian16) [inline pkg/helpers/helpers.go:423]: a unit  test?

File: pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go (hunk 3 above)
Comment (elgnay) [reply pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go]: It's good to delete non-terminating AppliedManifestWorks as well.

File: pkg/helpers/helpers.go (hunk 2 above)
Comment (elgnay) [reply pkg/helpers/helpers.go:423]: I moved the common part from klusterlet_status_controller.go to helpers.go.

File: pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go (hunk 1 above)
Comment (elgnay) [reply pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go]: Code refactored to get hub host from bootstrap secret.

File: pkg/helpers/helpers.go (hunk 2 above)
Comment (elgnay) [reply pkg/helpers/helpers.go:423]: Unit test case added.

File: pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go [hunk 4]
```diff
@@ -427,6 +459,56 @@ func (n *klusterletController) removeKlusterletFinalizer(ctx context.Context, de
...
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("unable to remove finalizer from AppliedManifestWork %q: %w", appliedManifestWork.Name, err)
+		}
+	}
+
+	// delete AppliedManifestWork if it does not start finalization yet
+	if appliedManifestWork.DeletionTimestamp.IsZero() {
```
Comment (qiujian16) [inline pkg/operators/klusterlet/controllers/klusterletcontroller/klusterlet_controller.go]: Do we need to delete appliedmanifestwork here?

**Pull request merged**